libyul/optimiser/ForLoopConditionIntoBody.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/ForLoopConditionIntoBody.h>
+#include <libyul/AsmData.h>
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+void ForLoopConditionIntoBody::operator()(Block& _block)
+{
+	for (Statement& _stmt: _block.statements)
+	{
+		if (_stmt.type() == typeid(ForLoop))
```

Comment:
Maybe it would make sense to skip this transform if the condition is already a literal constant. Then the transform would be idempotent.

Reply:
But then `ForLoopCondMoveOutBody(ForLoopCondIntoBody(forLoop))` won't be identity function anymore?

Reply:
Why not?

Reply:
Assume user inputs:
```
for {} 1 {}
{
  if a { break }
}
```
Then we first apply the skip transform version of ForLoopConditionIntoBody (which does nothing) then apply ForLoopConditionMoveOutBody, it will become
```
for {} iszero(a) {} {}
```

For some reason we may want optimization phases like `(ExpressionSplitter, ExpressionJoiner)`, `(ForLoopConditionIntoBody, ForLoopConditionMoveOutBody)` to be pairs of inverse functions? (ignoring the fact we may get `iszero(iszero(a))` after transformation)

Reply:
But isn't it a good thing that it will result in `for {} iszero(a) {} {}`?

Reply:
It may be better to think of such transforms not of performing a certain fixed transform but instead that the result of the transform satisfies certain syntactic or semantic conditions. And if that is the setting, then I would say transforms should take the "shortest path" towards satisfying that condition.

Reply:
That mindset makes sense! Will fix this tomorrow

Reply:
Updated. Was wondering if we should also skip this transformation if the iteration condition is Variable (which ExpressionSplitter also cannot apply on). But decided to keep it simple.